dandiapi/api/management/commands/extract_metadata.py

```diff
@@ -0,0 +1,121 @@
+    if asset.sha256 is not None:
+        return Digest(algorithm=DigestType.sha2_256, value=asset.sha256)
+
+    # Default to Etag
+    if asset.blob is not None:
+        return Digest(algorithm=DigestType.dandi_etag, value=asset.blob.etag)
+
+    raise Exception('Unsupported asset type')  # noqa: TRY002
+
+
+def extract_asset_metadata(asset: Asset, draft_version: Version):
+    readable_asset = RemoteReadableAsset(
+        asset.s3_url, size=asset.size, mtime=asset.modified, name=Path(asset.path).name
+    )
+
+    if not asset.path.lower().endswith('.nwb'):
+        logger.info('Asset %s: Not an NWB file, skipping...', asset.path)
+        return
+
+    new_metadata = nwb2asset(
+        readable_asset, digest=get_asset_digest(asset), schema_version=get_schema_version()
+    ).json_dict()
+
+    # Use dandiset owner, default to some admin user
+    user = (
+        draft_version.dandiset.owners.first()
+        or User.objects.filter(is_superuser=True, is_staff=True).first()
+    )
+
+    # Lock asset and version to ensure they don't change while performing this operation
+    with transaction.atomic():
+        locked_asset = Asset.objects.select_for_update().get(id=asset.id)
+        locked_version = Version.objects.select_for_update().get(id=draft_version.id)
+
+        # Ensure asset hasn't already been removed from the version
+        if not locked_version.assets.filter(id=locked_asset.id).exists():
+            raise Exception(f'Asset {locked_asset} no longer exists in version {locked_version}')  # noqa: TRY002
+
+        # Replace old asset with new asset containing updated metadata
+        change_asset(
+            user=user,
+            asset=locked_asset,
+            version=locked_version,
+            new_asset_blob=asset.blob,
+            new_zarr_archive=asset.zarr,
+            new_metadata=new_metadata,
+        )
+
+
+def extract_dandiset_assets(dandiset: Dandiset):
+    # Only update NWB assets which are out of date and do not belong to a published version
+    assets = dandiset.draft_version.assets.filter(
+        published=False,
+        path__iendswith='.nwb',
+        metadata__schemaVersion__lt=get_schema_version(),
+    ).select_related('blob', 'zarr')
+    if not assets:
+        logger.info('No old draft NWB assets found in dandiset %s. Skipping...', dandiset)
+        return
+
+    for asset in tqdm(assets):
+        extract_asset_metadata(asset=asset, draft_version=dandiset.draft_version)
+
+
+@group.command(help='Re-extracts the metadata of this asset')
+@click.argument('asset_id')
+def asset(asset_id: str):
+    asset = Asset.objects.get(asset_id=asset_id)
+    draft_versions = asset.versions.filter(version='draft')
+    if not draft_versions.exists():
+        raise click.ClickException(
+            'Cannot re-extract metadata of asset that has no associated draft versions.'
+        )
+
+    # Re-extract for every draft version
+    for version in draft_versions:
```

Comment:
This isn't likely necessary, but technically the number of versions is unbounded and we're running inside a management command, so there's not really any downsides
```suggestion
    for version in draft_versions.iterator():
```